### PR TITLE
Fix CUDA codegen crash for getElementPtr on rate-qualified vector types

### LINKED ISSUE #10417: [Regression] Segfault with backward pass in DiffTensorView
# Issue Description

`bwd_diff` through a `[Differentiable]` extension method on `DiffTensorView` that uses a `[ForceUnroll]` loop with index mutation (`x[R] += L`) produces a segfault at runtime when the method calls another `[Differentiable]` extension method in the loop body.

Specifically, `storeMatrixOnce` (loops over rows calling `storeVecOnce` with `x[R] += L`) and `loadMatrixOnce` (loops over rows calling `loadVecOnce` with `x[R] += L`) crash in the backward pass. The manually unrolled equivalent (calling `storeVecOnce`/`loadVecOnce` directly with fixed offsets) works correctly.

This is a regression: it works in slangc 2026.2.1 and segfaults in slangc 2026.3.1.

# Reproducer Code

**repro.slang:**

```slang
// Extension methods on DiffTensorView

__generic<T : __BuiltinFloatingPointType>
extension DiffTensorView<T> {

    // Vector load/store — backward works fine
    [Differentiable]
    __generic<let M : int, let R : int, let N : int>
    vector<T, M> loadVecOnce(vector<uint, N> x) {
        vector<T, M> result;
        [ForceUnroll]
        for (int j = 0; j < M; j++) {
            result[j] = this.loadOnce(x);
            x[R] += 1;
        }
        return result;
    }

    [Differentiable]
    __generic<let M : int, let N : int>
    vector<T, M> loadVecOnce(vector<uint, N> x) {
        return this.loadVecOnce<M, N - 1, N>(x);
    }

    [Differentiable]
    __generic<let M : int, let R : int, let N : int>
    void storeVecOnce(vector<uint, N> x, vector<T, M> y) {
        [ForceUnroll]
        for (int j = 0; j < M; j++) {
            this.storeOnce(x, y[j]);
            x[R] += 1;
        }
    }

    [Differentiable]
    __generic<let M : int, let N : int>
    void storeVecOnce(vector<uint, N> x, vector<T, M> y) {
        this.storeVecOnce<M, N - 1, N>(x, y);
    }

    // Matrix store — backward SEGFAULTS in 2026.3.1
    [Differentiable]
    __generic<let K : int, let L : int, let R : int, let N : int>
    void storeMatrixOnce(vector<uint, N> x, matrix<T, K, L> value) {
        [ForceUnroll]
        for (int i = 0; i < K; i++) {
            this.storeVecOnce<L, R, N>(x, value[i]);
            x[R] += L;
        }
    }

    [Differentiable]
    __generic<let K : int, let L : int, let N : int>
    void storeMatrixOnce(vector<uint, N> x, matrix<T, K, L> value) {
        this.storeMatrixOnce<K, L, N - 1, N>(x, value);
    }

    // Matrix load — backward SEGFAULTS in 2026.3.1
    [Differentiable]
    __generic<let K : int, let L : int, let R : int, let N : int>
    matrix<T, K, L> loadMatrixOnce(vector<uint, N> x) {
        matrix<T, K, L> result;
        [ForceUnroll]
        for (int i = 0; i < K; i++) {
            result[i] = this.loadVecOnce<L, R, N>(x);
            x[R] += L;
        }
        return result;
    }

    [Differentiable]
    __generic<let K : int, let L : int, let N : int>
    matrix<T, K, L> loadMatrixOnce(vector<uint, N> x) {
        return this.loadMatrixOnce<K, L, N - 1, N>(x);
    }
};

// SEGFAULTS in backward with slangc 2026.3.1:
[CUDAKernel]
[Differentiable]
[AutoPyBindCUDA]
void store_matrix_kernel(
    DiffTensorView translation, // (N, 3)
    DiffTensorView rotation,    // (N, 4)
    DiffTensorView result       // (N, 16)
) {
    uint idx = cudaThreadIdx().x + cudaBlockIdx().x * cudaBlockDim().x;
    if (idx >= translation.size(0)) return;

    float3 trans = translation.loadVecOnce<3>(uint2(idx, 0));
    float4 rot = rotation.loadVecOnce<4>(uint2(idx, 0));

    float4x4 mat;
    mat[0] = float4(rot.x, rot.y, rot.z, 0.0);
    mat[1] = float4(rot.y, rot.w, rot.x, 0.0);
    mat[2] = float4(rot.z, rot.x, rot.w, 0.0);
    mat[3] = float4(trans.x, trans.y, trans.z, 1.0);

    result.storeMatrixOnce(uint2(idx, 0), mat);  // <-- backward segfaults
}

// WORKS in backward — same logic, manually unrolled:
[CUDAKernel]
[Differentiable]
[AutoPyBindCUDA]
void store_matrix_manual_kernel(
    DiffTensorView translation, // (N, 3)
    DiffTensorView rotation,    // (N, 4)
    DiffTensorView result       // (N, 16)
) {
    uint idx = cudaThreadIdx().x + cudaBlockIdx().x * cudaBlockDim().x;
    if (idx >= translation.size(0)) return;

    float3 trans = translation.loadVecOnce<3>(uint2(idx, 0));
    float4 rot = rotation.loadVecOnce<4>(uint2(idx, 0));

    float4x4 mat;
    mat[0] = float4(rot.x, rot.y, rot.z, 0.0);
    mat[1] = float4(rot.y, rot.w, rot.x, 0.0);
    mat[2] = float4(rot.z, rot.x, rot.w, 0.0);
    mat[3] = float4(trans.x, trans.y, trans.z, 1.0);

    // Manually unrolled — backward works fine:
    result.storeVecOnce(uint2(idx, 0), mat[0]);
    result.storeVecOnce(uint2(idx, 4), mat[1]);
    result.storeVecOnce(uint2(idx, 8), mat[2]);
    result.storeVecOnce(uint2(idx, 12), mat[3]);
}
```

**repro.py** (requires `slangpy` and `torch`):

```python
import os
import torch
import slangpy as spy
import slangpy.torchintegration.torchtensormarshall  # registers torch.Tensor handler
from slangpy.torchintegration import diff_pair

torch.cuda.init()
handles = spy.get_cuda_current_context_native_handles()
device = spy.Device(
    type=spy.DeviceType.cuda,
    compiler_options={
        "include_paths": [],
        "defines": {"SLANGPY_DISPATCH": "1"},
        "optimization": spy.SlangOptimizationLevel.maximal,
        "floating_point_mode": spy.SlangFloatingPointMode.fast,
    },
    enable_print=True,
    enable_hot_reload=False,
    existing_device_handles=handles,
)

slang_file = os.path.join(os.path.dirname(os.path.abspath(__file__)), "repro.slang")
module = spy.Module.load_from_file(device, slang_file)

N = 4
t = torch.randn(N, 3, device="cuda")
r = torch.randn(N, 4, device="cuda")

# Test 1: manually unrolled — PASSES
print("Test 1: store_matrix_manual_kernel backward...")
res = torch.zeros(N, 16, device="cuda")
module.store_matrix_manual_kernel(_thread_count=N, translation=t, rotation=r, result=res)
gt, gr, gres = torch.zeros_like(t), torch.zeros_like(r), torch.ones_like(res)
module.store_matrix_manual_kernel.bwds(
    _thread_count=N,
    translation=diff_pair(t, gt), rotation=diff_pair(r, gr), result=diff_pair(res, gres))
print(f"  PASSED (grad={gt[0].tolist()})")

# Test 2: storeMatrixOnce — SEGFAULTS with slangc 2026.3.1
print("Test 2: store_matrix_kernel backward (storeMatrixOnce)...")
res2 = torch.zeros(N, 16, device="cuda")
module.store_matrix_kernel(_thread_count=N, translation=t.clone(), rotation=r.clone(), result=res2)
gt2, gr2, gres2 = torch.zeros_like(t), torch.zeros_like(r), torch.ones_like(res2)
print("  Running backward...", flush=True)
module.store_matrix_kernel.bwds(
    _thread_count=N,
    translation=diff_pair(t.clone(), gt2), rotation=diff_pair(r.clone(), gr2),
    result=diff_pair(res2, gres2))
print(f"  PASSED (grad={gt2[0].tolist()})")
```

# Expected Behavior

Both `store_matrix_kernel` and `store_matrix_manual_kernel` should produce correct gradients in the backward pass. `storeMatrixOnce` is semantically equivalent to manually calling `storeVecOnce` four times with fixed offsets — both should work identically.

# Actual Behavior

With slangc 2026.3.1, calling `.bwds()` on `store_matrix_kernel` (which uses `storeMatrixOnce`) produces a segfault:

```
Fatal Python error: Segmentation fault

Current thread 0x00007847cbfff6c0 (most recent call first):
  File "slangpy/core/function.py", line 287 in call    # _native_call
```

The manually unrolled `store_matrix_manual_kernel` works correctly with the same compiler version.

With slangc 2026.2.1, both kernels work correctly — no segfault.

# Environment

- **Slang Version**: Regression between 2026.2.1 (works) and 2026.3.1 (segfaults)
- **OS**: Linux (Ubuntu), kernel 6.14.0-35-generic, x86_64
- **slangpy version**: 0.40.1
- **PyTorch**: 2.7.0
- **Python**: 3.11

### PR DESCRIPTION
**Pull request by jhelferty-nv**
The `eliminatePhis` pass converts SSA `updateElement` instructions into
mutable variable operations (var + getElementPtr + store), preserving
`@ConstExpr` rate qualification from the original instruction. This
wraps the vector value type in `RateQualified(ConstExpr, ...)`.

The CPP/CUDA emitter's `getElementPtr` handler checks
`as<IRVectorType>(valueType)` to decide whether to emit `->x`/`->y`
accessors or `_slang_vector_get_element_ptr()`. When the value type is
rate-qualified, this cast fails, falling through to the generic emitter
which produces `vec[int(N)]` — invalid CUDA since vector types like
`uint2` don't support `operator[]`.

Fix by calling `unwrapAttributedType()` before the vector type check in
both `slang-emit-cpp.cpp` and `slang-emit-c-like.cpp`.

Added `tests/autodiff/nested-force-unroll-index-mutation.slang` which
reproduces the issue: nested `[ForceUnroll]` loops with `uint2` index
mutation inside differentiable functions, verified on CUDA target.

Fixes #10417

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by jhelferty-nv** [comment]
@saipraveenb25 Can you review this? The root cause analysis is on #10417.
